### DISCUSSION TITLE
Does Alpine need to use utils as global variables

### DISCUSSION THREAD

**Discussion by ericpugh** [alpinejs/alpine#2512] [4. General]
I've come across an issue with utilities from Alpine that are attached to `window` conflicting with other libraries. Specifically, I'm trying to use Alpine in a Drupal theme.  Drupal's most recent version (9.3) adds a standalone [library](https://www.npmjs.com/package/@drupal/once)  which adds `window.once()` as a global  method. The name conflict with Alpine's `window.once()` causes some pretty major problems. (Basically, all of Drupal core's ajax functionality is broken). (There's a discussion on [drupal.org](https://www.drupal.org/project/drupal/issues/3254840#comment-14343697))

I'm not familiar enough with Alpine's internals, is there a technical reason that Alpine needs to add the directive utils to the global scope? Would it be possible to import these scoped locally or alias something less likely to conflict?

**Comment by SimoTod** [comment]
Are you building your own bundle? The official cdn version doesn't expose anything on the window object. (everything should be wrapped in an IIFE)

**Reply by ericpugh** [reply]
Yes, I'm adding some custom magic functions and whatnot. It's true this is somewhat an edge case since it won't happen with the CDN library and you can avoid it by setting up your build a certain way.

But just wondering if it was worth avoiding this possibility (the developers on the Drupal issue are having a similar discussion)

**Reply by SimoTod** [reply]
I assume it depends on the bundler, Alpine doesn't set those functions on the global namespace, it just exports them from utils.js and esbuild generates a valid output. 

It's probably fine if you use the compiled js as a module. I haven't tried but it's also probably ok if you wrap your import/start bit in a function scope.

```
function() {
  ... code here ... 
}()
```

**Reply by ericpugh** [reply]
Oh. You're right. So just... "build it as an IIFE". I feel pretty silly now. :smile: Thanks for talking me through it!

Maybe, a section in the Docs under "Extending" on [build](https://github.com/alpinejs/alpine/blob/main/scripts/build.js) would help some people? You know, just something simple like, this is how Alpine is built, this is what you it would look like if you did a custom build but still wanted it to work with the CDN plugins (or some simple scenario).

**Reply by SimoTod** [reply]
It sounds sensible. The docs is in the repo (package > docs) if you want to send a PR 👍
